File: LC_238.py
from typing import *

class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        res = [1]*len(nums)

        #[1, 2, 3]
        #[1, 1, 1]
        pre = 1
        for i in range(len(nums)):
            res[i] = pre
            pre *= nums[i] 
        #[1, 2, 6]
        
        post = 1
        for i in range(len(nums)-1, -1, -1):
             res[i] *= post #multiply the right element, nothing here so 1
             post *= nums[i]    #multiply the current element so that we can store the value 
        print(res)
        return res
    
Solution().productExceptSelf([1, 2, 3])

File: LC_238.java
import java.util.Arrays;

class Solution {
    public static int[] productExceptSelf(int[] nums) {
        int [] res = new int[nums.length];

        Arrays.fill(res, 1);

        int left = 1;

        for (int i = 0; i < res.length; i++) {
            res[i] = left;
            left *= nums[i];
        }

        //[1, 1, 2, 6]

        int right = 1;
        for (int i = res.length - 1; i >= 0; i--) {
            res[i] *= right;
            right *= nums[i];
        }


        return res;
    }

    public static void main(String[] args) {
        System.out.println( Arrays.toString(productExceptSelf(new int[]{1, 2, 4, 5})));
    }
}